Add 348 release notes (#6258)
prestosql/presto: presto-docs/src/main/sphinx/release/release-348.md

# Release 348 (XX Dec 2020)

## General Changes

* Add support for `DISTINCT` clause in aggregations within correlated subqueries. ({issue}`5904`)
* Improve query performance by reducing worker to worker communication overhead. ({issue}`6126`)
* Reduce memory pressure and improve performance of queries that contain join. ({issue}`6176`)
* Fix ``EXPLAIN ANALYZE`` for certain queries that contain broadcast join. ({issue}`6115`)

> * Fix planning failures for queries that contain filtered aggregations and outer joins. ({issue}`6141`)

Reviewer: What are "filtered aggregations"?

Reviewer 2: https://prestosql.io/docs/current/functions/aggregate.html#filtering-during-aggregation

Reviewer: Thanks. Maybe we should quote the syntax in the release notes:
```
Fix planning failures for queries that contain outer joins and aggregations using `FILTER (WHERE <condition>)` syntax.
```

## JDBC Driver Changes

* Allow reading `timestamp with time zone` value as `ZonedDateTime` using `ResultSet.getObject(int column, Class<?> type)` method. ({issue}`307`)
* Change representation of a `row` value. `ResultSet.getObject` now returns an instance of `io.prestosql.jdbc.Row` class, which better represents
  the returned value. Previously a `row` value was represented as a `Map` instance, with unnamed fields being named like `field0`, `field1`, etc. 
  You can access the previous behavior by invoking `getObject(column, Map.class)` on the `ResultSet` object. ({issue}`4588`)
* Fix failure when reading a `timestamp` or `timestamp with time zone` value with second fraction greater than or equal to 999999999500 picoseconds. ({issue}`6147`)
* Fix element representation in arrays returned from `ResultSet.getArray`, making it consistent with `ResultSet.getObject`.
  Previously the elements were represented using internal client representation (e.g. `String`). ({issue}`6048`)
* Fix `ResultSetMetaData.getColumnType` for `timestamp with time zone`. Previously the type was miscategorized as `java.sql.Types.TIMESTAMP`. ({issue}`6251`)

> * Fix failure when instance of `SphericalGeography` Geospatial type is returned in the `ResultSet`. ({issue}`6240`)

Reviewer 3: @findepi Did it happen always? What was returned previously?

* Fix query progress reporting. ({issue}`6119`)
* Fix failure when instance of `SphericalGeography` Geospatial type is returned to the client. ({issue}`6238`)

## Hive Connector Changes

* Reduce scheduling latency for queries where scanned files that are referenced by symlinks
  are located in same directory. ({issue}`6158`)
* Add deserializer class name to hive connector split information, which is exposed through the event listener. ({issue}`6006`)
* Improve parallelism for queries where scanned files are referenced by symlinks. ({issue}`6213`)

## Iceberg Connector Changes

* Optimize predicate pushdown by using filters on non-partition columns during split generation and table scan. ({issue}`4932`)

## Kafka Connector Changes

* Allow writing `timestamp with time zone` values into columns using `milliseconds-since-epoch` or `seconds-since-epoch` JSON encoders. ({issue}`6074`)

## Other Connector Changes

* Fix table metadata information caching for PostgreSQL, MySQL, SQL Server, Redshift, MemSQL and Phoenix connectors. Previously cache did not 

>   store information if table exists and it was always verifing that in remote database.  ({issue}`6081`)

Reviewer 3: Maybe
`Previously cache did not keep information whether table exists and remote database was always queried for table existence`

# Release 348 (XX Dec 2020)

## General Changes

* Add support for `DISTINCT` clause in aggregations within correlated subqueries. ({issue}`5904`)
* Improve query performance by reducing worker to worker communication overhead. ({issue}`6126`)
* Reduce memory pressure and improve performance of queries that contain join. ({issue}`6176`)
* Fix ``EXPLAIN ANALYZE`` for certain queries that contain broadcast join. ({issue}`6115`)
* Fix planning failures for queries that contain filtered aggregations and outer joins. ({issue}`6141`)
* Fix incorrect results when correlated subquery in join contains aggregation functions such as `array_agg` or `checksum`. ({issue}`6145`)
* Fix duplicate call to event listener on query completion, if query failed early during preparation. ({issue}`6103`)

## Web UI

* Add support for OAuth2 authorization in Web UI. ({issue}`5355`)

> * Fix invalid operator stats reporting in stage performance view. ({issue}`6114`)

Reviewer: `s/reporting/being reported/`?

## JDBC Driver Changes

* Allow reading `timestamp with time zone` value as `ZonedDateTime` using `ResultSet.getObject(int column, Class<?> type)` method. ({issue}`307`)
* Change representation of a `row` value. `ResultSet.getObject` now returns an instance of `io.prestosql.jdbc.Row` class, which better represents
  the returned value. Previously a `row` value was represented as a `Map` instance, with unnamed fields being named like `field0`, `field1`, etc. 
  You can access the previous behavior by invoking `getObject(column, Map.class)` on the `ResultSet` object. ({issue}`4588`)

> * Fix failure when reading a `timestamp` or `timestamp with time zone` value with second fraction greater than or equal to 999999999500 picoseconds. ({issue}`6147`)

Reviewer: ```suggestion
* Fix failure when reading a `timestamp` or `timestamp with time zone` value with a second fraction greater than or equal to 999999999500 picoseconds. ({issue}`6147`)
```
I am not sure though

Author: Actually, "with seconds fraction" sounds better.

* Fix element representation in arrays returned from `ResultSet.getArray`, making it consistent with `ResultSet.getObject`.
  Previously the elements were represented using internal client representation (e.g. `String`). ({issue}`6048`)
* Fix `ResultSetMetaData.getColumnType` for `timestamp with time zone`. Previously the type was miscategorized as `java.sql.Types.TIMESTAMP`. ({issue}`6251`)

> * Fix failure when instance of `SphericalGeography` Geospatial type is returned in the `ResultSet`. ({issue}`6240`)

Reviewer: ```suggestion
* Fix failure when an instance of `SphericalGeography` Geospatial type is returned in the `ResultSet`. ({issue}`6240`)
```

## CLI Changes

* Fix rendering of `row` values with unnamed fields. Previously they were printed using fake field names like `field0`, `field1`, etc. ({issue}`4587`)
* Fix query progress reporting. ({issue}`6119`)

> * Fix failure when instance of `SphericalGeography` Geospatial type is returned to the client. ({issue}`6238`)

Reviewer: ```suggestion
* Fix failure when an instance of `SphericalGeography` Geospatial type is returned to the client. ({issue}`6238`)
```

## Hive Connector Changes

> * Reduce scheduling latency for queries where scanned files that are referenced by symlinks

Reviewer: I feel some commas missing in this sentence. native-english-speaker needed :)

Reviewer 4: I am not sure what this actually means so I cant reword.. can you maybe explain?

>   are located in same directory. ({issue}`6158`)

Reviewer: ```suggestion
  are located in the same directory. ({issue}`6158`)
```

* Add deserializer class name to hive connector split information, which is exposed through the event listener. ({issue}`6006`)
* Improve parallelism for queries where scanned files are referenced by symlinks. ({issue}`6213`)

## Iceberg Connector Changes

* Optimize predicate pushdown by using filters on non-partition columns during split generation and table scan. ({issue}`4932`)

## Kafka Connector Changes

* Allow writing `timestamp with time zone` values into columns using `milliseconds-since-epoch` or `seconds-since-epoch` JSON encoders. ({issue}`6074`)

## Other Connector Changes

* Fix table metadata caching for PostgreSQL, MySQL, SQL Server, Redshift, MemSQL and Phoenix connectors:
  * Previously cache did not store information if table exists and it was always verifing that in remote database.  ({issue}`6081`)

>   * Consider session properties when caching metadata, so there are different cache entries for diffent session properties. ({issue}`6167`)

Reviewer 5: Can this be described more from user perspective?

This is actually "Fix potential query failures when using caching" situation but the wording doesn't apparently communicate this.

>   * Fix cache invalidation to remove entries for all users not only for the user that is running the query. ({issue}`6167`)

Reviewer 2: Is this the way to do nested indentation?

Reviewer 6: How about 
```
Fix cache invalidation by removing entries for all users, not only for the user running the query. ({issue}`6167`)
```

Reviewer 4: Since the top level item is the only one .. just make that a separate sentence and then use the follow up ones as normal list

Reviewer 5: @phd3 " Fix cache invalidation ..." doesn't describe the problem for the user. Is it perf improvement (better caching) or correctness, or robustness fix?

>   * Previously cache did not store information if table exists and it was always verifing that in remote database.  ({issue}`6081`)
>   * Consider session properties when caching metadata, so there are different cache entries for diffent session properties. ({issue}`6167`)
>   * Fix cache invalidation to remove entries for all users not only for the user that is running the query. ({issue}`6167`)

Reviewer: ```suggestion
  * Previously cache did not store information if table exists and it was always verifying that in the remote database.  ({issue}`6081`)
  * Consider session properties when caching metadata, so there are different cache entries for different session properties. ({issue}`6167`)
  * Fix cache invalidation to remove entries for all users not only for the user that is running the query. ({issue}`6167`)
```

>   * Fix cache invalidation to remove entries for all users not only for the user that is running the query. ({issue}`6167`)

Reviewer 5: Can this be described more from user perspective?

This is actually "Fix potential query failures -- eg SELECT * seeing a column that has already been removed -- when using caching" situation but the wording doesn't apparently communicate this.

## SPI Changes

* Change `SystemAccessControl#filterColumns` and `ConnectorAccessControl#filterColumns` method to accept a set of 

>   column names and return a set of visible column names. ({issue}`6084`))

Reviewer 7: Remove redundant `))` → `)`

# Release 348 (XX Dec 2020)

## General Changes

* Add support for `DISTINCT` clause in aggregations within correlated subqueries. ({issue}`5904`)
* Improve query performance by reducing worker to worker communication overhead. ({issue}`6126`)

> * Reduce memory pressure and improve performance of queries that contain join. ({issue}`6176`)

Reviewer 4: join statements ?

* Fix ``EXPLAIN ANALYZE`` for certain queries that contain broadcast join. ({issue}`6115`)
* Fix planning failures for queries that contain outer joins and aggregations using `FILTER (WHERE <condition>)` syntax. ({issue}`6141`)
* Fix incorrect results when correlated subquery in join contains aggregation functions such as `array_agg` or `checksum`. ({issue}`6145`)
* Fix duplicate call to event listener on query completion, if query failed early during preparation. ({issue}`6103`)

> 

Reviewer 8: Add:
```
* Improve performance of top-n queries. ({issue}`6072`)
```

## Web UI

> * Add support for OAuth2 authorization in Web UI. ({issue}`5355`)

Reviewer 4: remove redundant "in Web UI"

* Fix invalid operator stats being reported in stage performance view. ({issue}`6114`)

## JDBC Driver Changes

* Allow reading `timestamp with time zone` value as `ZonedDateTime` using `ResultSet.getObject(int column, Class<?> type)` method. ({issue}`307`)

> * Change representation of a `row` value. `ResultSet.getObject` now returns an instance of `io.prestosql.jdbc.Row` class, which better represents

Reviewer 4: Is this a breaking change for users of the JDBC driver?

Reviewer: I think it is reasonable to mark the bullet with something like `Warning: incompatible change` or `Warning: behavior change`. Not sure what wording should be used here.

Reviewer: For the future: maybe we should have some graphical representation to mark notes which introduce backward incompatible changes?

  the returned value. Previously a `row` value was represented as a `Map` instance, with unnamed fields being named like `field0`, `field1`, etc. 
  You can access the previous behavior by invoking `getObject(column, Map.class)` on the `ResultSet` object. ({issue}`4588`)
* Represent `varbinary` value using hex string representation in `ResultSet.getString`. Previously the return value was useless, similar to `"B@2de82bf8"`. ({issue}`6247`)
* Consider session properties when caching metadata. ({issue}`6167`)
* Fix failure when reading a `timestamp` or `timestamp with time zone` value with seconds fraction greater than or equal to 999999999500 picoseconds. ({issue}`6147`)
* Fix failure when reading a `time` value with seconds fraction greater than or equal to 999999999500 picoseconds. ({issue}`6204`)

> * Fix element representation in arrays returned from `ResultSet.getArray`, making it consistent with `ResultSet.getObject`.

Reviewer 4: Again.. do we need to call this out as breaking change?

Reviewer: Same here.

  Previously the elements were represented using internal client representation (e.g. `String`). ({issue}`6048`)

> * Fix `ResultSetMetaData.getColumnType` for `timestamp with time zone`. Previously the type was miscategorized as `java.sql.Types.TIMESTAMP`. ({issue}`6251`)

Reviewer 4: Should we add what it does now?

Reviewer: Yeah I guess.
Maybe instead sth like this:
```
`ResultSetMetaData.getColumnType`now returns `java.sql.Types.TIMESTAMP_WITH_TIMEZONE` for `timestamp with time zone`. Previously the type was miscategorized as `java.sql.Types.TIMESTAMP`.

Reviewer: @findepi should we add similar entry for `time with time zone`?

Reviewer 5: Yes, both shoulkd be covered

Reviewer 5: @martint can you please add similar one for `time with time zone`?
sorry for missing it in the RN issue

> * Fix failure when an instance of `SphericalGeography` Geospatial type is returned in the `ResultSet`. ({issue}`6240`)

Reviewer 4: lowercase geospatial

* Fix cache entries invalidation, so entries are removed for all users, not only the one that was running the query. ({issue}`6167`)

## CLI Changes

> * Fix rendering of `row` values with unnamed fields. Previously they were printed using fake field names like `field0`, `field1`, etc. ({issue}`4587`)

Reviewer: Or just say. `Drop rendering of fake field names for `row` values with unnamed fields`.

* Fix query progress reporting. ({issue}`6119`)

> * Fix failure when an instance of `SphericalGeography` Geospatial type is returned to the client. ({issue}`6238`)

Reviewer 4: lowercase geospatial

## Hive Connector Changes

* Reduce scheduling latency for queries where scanned files that are referenced by symlinks
  are located in the same directory. ({issue}`6158`)

> * Add deserializer class name to split information, which is exposed through the event listener. ({issue}`6006`)

Reviewer 6: the suggestion looks good.

> * Add deserializer class name to hive connector split information, which is exposed through the event listener. ({issue}`6006`)

Reviewer 4: Uppercase "Hive connector"

* Improve parallelism for queries where scanned files are referenced by symlinks. ({issue}`6213`)
* Add support for S3 streaming uploads. Data is uploaded to S3 as it is written, rather

>   than staging to a local temporary file. This feature is disabled by default and can be enabled

Reviewer 4: rather than staged

Reviewer 4: This feature is disabled by default, and

>   using the `hive.s3.streaming.enabled` configuration property. ({issue}`3712`, {issue}`6201`)

Reviewer 4: also .. is this property documented in https://prestosql.io/docs/current/connector/hive-s3.html ?

* Decrease default refresh thread count for metastore cache. ({issue}`6156`)

> * Reduce load on metastore when metastore caching is enabled with background refresh

Reviewer 4: Break up into two sentenced maybe?

Reviewer 8: maybe?
```
Reduce load on metastore when background cache refresh is enabled by not refreshing partition metadata and statistics in the background.
```

  by removing background refresh for partition metadata and statistics. ({issue}`6101`)
* Verify that data is in the correct bucket file when reading bucketed tables.
  This is enabled by default, as incorrect bucketing can cause incorrect query results,

>   but can be disabled using the `hive.validate-bucketing` configuration property

Reviewer 4: is this documented in hive.rst ?

Reviewer: nope

  or the `validate_bucketing` session property. ({issue}`6012`)
* Allow configuring S3 endpoint in security mapping. ({issue}`3869`)
* Allow fallback to legacy Hive view translation logic via `hive.legacy-hive-view-translation` config property or 
  `legacy_hive_view_translation` session property. ({issue}`6195 `)
    
## Iceberg Connector Changes

> * Optimize predicate pushdown by using filters on non-partition columns during split generation and table scan. ({issue}`4932`)

Reviewer: Can we express this more in terms of user visible effects? Is a broader set of expressions pushed down?  Can we classify those somehow here?

Author: @phd3

Reviewer 6: @losipiuk How about the following? Or may be even just the first sentence of it?

```
Improve performance of queries containing filters on non-partition columns. Such filters are now used for optimizing  split generation and table scan. 
```

Reviewer: Looks good to me. Thanks.

> * Remove extra file system stat call when opening files. ({issue}`6174`)

Reviewer: What is the user visible effect? Performance improvement? Maybe we should drop that at all if it is negligible.

Author: @electrum

Reviewer 8: It is a performance improvement, but I believe it is not much. Maybe drop this entirelly?

* Add support for Google Cloud Storage and Azure Storage. ({issue}`6186`)

## Kafka Connector Changes

* Allow writing `timestamp with time zone` values into columns using `milliseconds-since-epoch` or `seconds-since-epoch` JSON encoders. ({issue}`6074`)

## Other Connector Changes

* Fix table metadata caching for PostgreSQL, MySQL, SQL Server, Redshift, MemSQL and Phoenix connectors:

>   * Previously cache did not store information if table exists and it was always verifying that in the remote database. ({issue}`6081`)

Reviewer 4: if tables exists, and

>   * Consider session properties when caching metadata, so there are different cache entries for different session properties. ({issue}`6167`)

Reviewer 8: `Session properties are now considered when caching metadata`

  * Fix cache invalidation to remove entries for all users not only for the user that is running the query. ({issue}`6167`)

> ## SPI Changes

Reviewer 9: Can we move SPI changes after `General Changes` ? Sometime it might be skipped.

Reviewer 5: I think we usally have them at the bottom

Reviewer: I think those should be at the bottom. Most consumers of Presto do not care much about those.

>   * Fix cache invalidation to remove entries for all users not only for the user that is running the query. ({issue}`6167`)

Reviewer 4: comma: for all users, not only for the user...

* Change `SystemAccessControl#filterColumns` and `ConnectorAccessControl#filterColumns` method to accept a set of 

>   column names and return a set of visible column names. ({issue}`6084`)

Reviewer 4: comma: column names, and return..

# Release 348 (XX Dec 2020)

## General Changes

* Add support for `DISTINCT` clause in aggregations within correlated subqueries. ({issue}`5904`)
* Improve query performance by reducing worker to worker communication overhead. ({issue}`6126`)
* Reduce memory pressure and improve performance of queries that contain joins. ({issue}`6176`)
* Fix ``EXPLAIN ANALYZE`` for certain queries that contain broadcast join. ({issue}`6115`)
* Fix planning failures for queries that contain outer joins and aggregations using `FILTER (WHERE <condition>)` syntax. ({issue}`6141`)
* Fix incorrect results when correlated subquery in join contains aggregation functions such as `array_agg` or `checksum`. ({issue}`6145`)
* Fix duplicate call to event listener on query completion, if query failed early during preparation. ({issue}`6103`)
* Support arbitrary query in `SHOW STATS`. ({issue}`3109`)

> * Output of `SHOW STATS` is changed ({issue}`3109`): 

Reviewer 2: I am not sure if this is too detailed entry. Maybe we should just remove it?

Author: Yes, let's remove it.

* Fix ``EXPLAIN ANALYZE`` for certain queries that contain broadcast join. ({issue}`6115`)
* Fix planning failures for queries that contain outer joins and aggregations using `FILTER (WHERE <condition>)` syntax. ({issue}`6141`)
* Fix incorrect results when correlated subquery in join contains aggregation functions such as `array_agg` or `checksum`. ({issue}`6145`)
* Fix incorrect query results when using `timestamp with time zone` constants with precision higher than 3 
  describing same point in time but in different zones. ({issue}`6318`)
* Fix duplicate query completion events if query fails early. ({issue}`6103`)
* Fix query failure when views are accessed and current session does not
  specify default schema and catalog. ({issue}`6294`)

## Web UI

* Add support for OAuth2 authorization. ({issue}`5355`)
* Fix invalid operator stats in Stage Performance view. ({issue}`6114`)

## JDBC Driver Changes

* Allow reading `timestamp with time zone` value as `ZonedDateTime` using `ResultSet.getObject(int column, Class<?> type)` method. ({issue}`307`)
* Accept `java.time.LocalDate` in `PreparedStatement.setObject(int, Object)`. ({issue}`6301`)

> * Extend `PreparedStatement.setObject(int, Object, int)` to allow setting `time` and `timestamp` values with precision higher than nanoseconds. 

Reviewer: This description is long and still does not pass all the information needed to do the actual call (what value of `java.sql.Types` should be used).
I think we should either make it shorter; like:
```
Extend `PreparedStatement.setObject(int, Object, int)` to allow setting `time` and `timestamp` values with precision higher than nanoseconds. 
This can be done via `String` object representing valid SQL literal.
```
Or extend the `alongside...` part.

cc: @findepi

* Fix `ResultSetMetaData.getColumnType` for `timestamp with time zone`. Previously the type was miscategorized as `java.sql.Types.TIMESTAMP`. ({issue}`6251`)
* Fix `ResultSetMetaData.getColumnType` for `time with time zone`. Previously the type was miscategorized as `java.sql.Types.TIME`. ({issue}`6251`)
* Fix failure when an instance of `SphericalGeography` geospatial type is returned in the `ResultSet`. ({issue}`6240`)
  
## CLI Changes

* Fix rendering of `row` values with unnamed fields. Previously they were printed using fake field names like `field0`, `field1`, etc. ({issue}`4587`)
* Fix query progress reporting. ({issue}`6119`)
* Fix failure when an instance of `SphericalGeography` geospatial type is returned to the client. ({issue}`6238`)

## Hive Connector Changes

* Allow configuring S3 endpoint in security mapping. ({issue}`3869`)
* Add support for S3 streaming uploads. Data is uploaded to S3 as it is written, rather
  than staged to a local temporary file. This feature is disabled by default, and can be enabled
  using the `hive.s3.streaming.enabled` configuration property. ({issue}`3712`, {issue}`6201`)

> * Decrease default refresh thread count for metastore cache. ({issue}`6156`)

Reviewer: User visible effect?

> * Reduce load on metastore when background cache refresh is enabled. Partition metadata and statistics are 

Reviewer: It seems to me we should merge this point with one above and just use `Reduce load on metastore when background cache refresh is enabled` as description.

* Verify that data is in the correct bucket file when reading bucketed tables.
  This is enabled by default, as incorrect bucketing can cause incorrect query results,
  but can be disabled using the `hive.validate-bucketing` configuration property
  or the `validate_bucketing` session property. ({issue}`6012`)
* Allow fallback to legacy Hive view translation logic via `hive.legacy-hive-view-translation` config property or 
  `legacy_hive_view_translation` session property. ({issue}`6195 `)
* Add deserializer class name to split information exposed to the event listener. ({issue}`6006`)
* Improve performance when querying tables that contain symlinks. ({issue}`6158`, {issue}`6213`)
    
## Iceberg Connector Changes

* Improve performance of queries containing filters on non-partition columns. Such filters are now used 
  for optimizing split generation and table scan.  ({issue}`4932`)
* Add support for Google Cloud Storage and Azure Storage. ({issue}`6186`)

## Kafka Connector Changes

* Allow writing `timestamp with time zone` values into columns using `milliseconds-since-epoch` or 
  `seconds-since-epoch` JSON encoders. ({issue}`6074`)

## Other Connector Changes

> * Fix table metadata caching for PostgreSQL, MySQL, SQL Server, Redshift, MemSQL and Phoenix connectors. ({issue}`6081`, {issue}`6167`)

Reviewer: It would be nice to capture shortly what is the nature of fix? Was cache not working? Or misbehave? Can it be done using single sentence?

Author: I changed it to "Fix ineffective table metadata caching..."
